rollup/pipeline/pipeline.go

```diff
@@ -0,0 +1,412 @@
+	parent   *types.Block
+	start    time.Time
+
+	// accumalators
+	ccc            *circuitcapacitychecker.CircuitCapacityChecker
+	Header         types.Header
+	state          *state.StateDB
+	nextL1MsgIndex uint64
+	blockSize      common.StorageSize
+	txs            types.Transactions
+	coalescedLogs  []*types.Log
+	receipts       types.Receipts
+	gasPool        *core.GasPool
+
+	// com channels
+	TxnQueue         chan *types.Transaction
+	applyStageRespCh chan error
+	ResultCh         chan *Result
+
+	// Test hooks
+	beforeTxHook func() // Method to call before processing a transaction.
+}
+
+func NewPipeline(
+	chain *core.BlockChain,
+	vmConfig *vm.Config,
+	state *state.StateDB,
+	coinbase *common.Address,
+
+	header *types.Header,
+	ccc *circuitcapacitychecker.CircuitCapacityChecker,
+) *Pipeline {
+	parent := chain.GetBlock(header.ParentHash, header.Number.Uint64()-1)
+	return &Pipeline{
+		chain:          chain,
+		vmConfig:       vmConfig,
+		parent:         parent,
+		nextL1MsgIndex: *rawdb.ReadFirstQueueIndexNotInL2Block(chain.Database(), parent.Hash()),
+		Header:         *header,
+		ccc:            ccc,
+		coinbase:       coinbase,
+		state:          state,
+		gasPool:        new(core.GasPool).AddGas(header.GasLimit),
+	}
+}
+
+func (p *Pipeline) WithBeforeTxHook(beforeTxHook func()) *Pipeline {
+	p.beforeTxHook = beforeTxHook
+	return p
+}
+
+func (p *Pipeline) Start(deadline time.Time) error {
+	p.start = time.Now()
+	p.TxnQueue = make(chan *types.Transaction)
+	applyStageRespCh, incrementCh, err := p.traceAndApplyStage(p.TxnQueue)
+	if err != nil {
+		log.Error("Failed starting traceAndApplyStage", "err", err)
+		return err
+	}
+	p.applyStageRespCh = applyStageRespCh
+	p.ResultCh = p.cccStage(incrementCh, deadline)
+	return nil
+}
+
+func (p *Pipeline) TryPushTxns(txs types.OrderedTransactionSet, onFailingTxn func(txnIndex int, tx *types.Transaction, err error) bool) *Result {
+	for {
+		tx := txs.Peek()
+		if tx == nil {
+			break
+		}
+
+		result, err := p.TryPushTxn(tx)
+		if result != nil {
+			return result
+		}
+
+		switch {
+		case err == nil, errors.Is(err, core.ErrNonceTooLow):
+			txs.Shift()
+		default:
+			if errors.Is(err, ErrApplyStageDone) || onFailingTxn(p.txs.Len(), tx, err) {
+				close(p.TxnQueue)
+				p.TxnQueue = nil
+				return nil
+			}
+
+			if tx.IsL1MessageTx() {
+				txs.Shift()
+			} else {
+				txs.Pop()
+			}
+		}
+	}
+
+	return nil
+}
+
+func (p *Pipeline) TryPushTxn(tx *types.Transaction) (*Result, error) {
+	select {
+	case p.TxnQueue <- tx:
+	case res := <-p.ResultCh:
+		return res, nil
+	}
+
+	select {
+	case err, valid := <-p.applyStageRespCh:
+		if !valid {
+			return nil, ErrApplyStageDone
+		}
+		return nil, err
+	case res := <-p.ResultCh:
+		return res, nil
+	}
+}
+
+func (p *Pipeline) Kill() {
+	if p.TxnQueue != nil {
+		close(p.TxnQueue)
+	}
+
+	select {
+	case <-p.applyStageRespCh:
+		<-p.ResultCh
+	case <-p.ResultCh:
+		<-p.applyStageRespCh
+	}
+}
+
+type PendingBlockIncrement struct {
```

Comment:
This does not feel an "increment" to me, since `Header`, `State`, `Txs` etc contain all changes up to and including the latest tx processed (not just the incremental change). Not sure what might be a better name though.

Reply:
How about "progress"?

Reply:
How about `BlockCandidate` or `NewBlockCandidate`?

Reply:
`NewBlockCandidate`sounds good.

Reply:
done https://github.com/scroll-tech/go-ethereum/pull/735/commits/a53eb879b83847b86e021509a39d10321c9d38db